library/dqueue.js

```diff
@@ -0,0 +1,137 @@
+   * @param {integer} lengthOfDqueue
+   * @return {dqueue}
+   */
+  constructor(lengthOfDqueue) {
+    this.verifyLengthInput(lengthOfDqueue);
+    this.lengthOfDqueue = lengthOfDqueue;
+    this.head = 0;
+    this.tail = this.lengthOfDqueue - 1;
+    this.count = 0;
+    this.dqueue = new Array(this.lengthOfDqueue);
+  }
+
+  verifyLengthInput(lengthOfDqueue) {
+    /**
+     * Verify the lengthOfQueue provided is an integer.
+     * @param {integer} lengthOfDqueue
+     * @return {integer}
+     */
+    if (!Number.isInteger(lengthOfDqueue)) {
+      throw new Error("Only accept integer as length");
+    }
+    return lengthOfDqueue;
+  }
+
+  pushToFront(value) {
+    /**
+     * Pushing value onto the dqueue from front(clockwise), if full throw error
+     * adjusting the value of the head when it is equal to the length
+     * @param {any} value
+     * @return {array}
+     */
+    if (this.count === this.lengthOfDqueue) {
+      throw new Error("Error");
+    }
+    this.dqueue[this.head] = value;
+    this.count += 1;
+    this.head += 1;
+    this.head = this.head % this.lengthOfDqueue;
+    return this.dqueue;
+  }
+
+  pushToBack(value) {
+    /**
+     * Pushing value onto the dqueue from back(counter-clockwise), if full throw error
+     * adjusting the value of tail when it is negative
+     * @param {any} value
+     * @return {array}
+     */
+    if (this.count === this.lengthOfDqueue) {
+      throw new Error("Error");
+    }
+    this.dqueue[this.tail] = value;
+    this.count += 1;
+    this.tail -= 1;
+    this.tail = this.tail < 0 ? this.lengthOfDqueue - 1 : this.tail;
+    return this.dqueue;
+  }
+
+  popFromFront() {
+    /**
+     * Popping values from dqueue from front, if empty throw error
+     * Popping values from the front === last value added clockwise.
+     * Readjusting the value of head when it is negative
+     * When dqueue is empty, reset head and tail values.
+     * @return {any}
+     */
+    if (this.count === 0) {
+      throw new Error("Error");
+    }
+    this.head -= 1;
+    this.head =
+      this.head < 0 ? (this.head = this.lengthOfDqueue - 1) : this.head;
+    let poppedValue = this.dqueue[this.head];
+    this.dqueue[this.head] = null;
+    this.count -= 1;
+    this.resetValuesOnEmpty();
+    return poppedValue;
+  }
+
+  popFromBack() {
+    /**
+     * Popping values from dqueue from back, if empty throw error
+     * Reset tail value when it is equal to the length
+     * When dqueue is empty, reset head and tail values.
+     * Popping values from the back === first value added to dqueue when going counter- clockwise.
+     * @return {any}
+     */
+    if (this.count === 0) {
+      throw new Error("Error");
+    }
+    this.tail += 1;
+    this.tail = this.tail % this.lengthOfDqueue;
+    let poppedValue = this.dqueue[this.tail];
+    this.dqueue[this.tail] = null;
+    this.count -= 1;
+    this.resetValuesOnEmpty();
```

Comment:
why was this needed, won't it continue to work no matter where the pointers are ?

Reply:
Not needed, as the mod function will perform the resetting. however resetting is required, otherwise pointers will go out of range of length.